src/differentials.jl

```diff
@@ -206,3 +206,27 @@ end
 end
 
 Base.conj(x::Thunk) = @thunk(conj(extern(x)))
+
+
+####
+"""
+    NO_FIELDS
+
+Constant for the reverse-mode derivative with respect to a structure that has no fields.
+The most notable use for this is for the reverse-mode derivative with respect to the
+function itself, when that function is not a closure.
+"""
+const NO_FIELDS = DNE()
+
+####
+"""
+    differential(𝒟::Type, der)
+
+For some differential (e.g. a `Number`, `AbstractDifferential`, `Matrix`, etc.),
+convert it to another differential that is more suited for the domain given by
+the type 𝒟.
+"""
+function differential(::Type{<:Union{<:Real, AbstractArray{<:Real}}}, w::Wirtinger)
```

Comment:
Maybe check for `eltype` instead, to allow for iterables that are not `<: AbstractArray`?

Reply:
right now the idea of having other iterators as differentials is not something I've fully worked though in my head.

Reply:
Ok, this should be fine for now. There's definitely still a lot to do for supporting `Wirtinger` of arrays correctly.